### PR TITLE
Fix CUDAGraph stale output checks for cached tensors

### LINKED ISSUE #122192: extra checks for reduce-overhead inconsistent when warmup/recording
### 🐛 Describe the bug

Running
```
import torch
cfn = torch.compile(torch.sin, mode='reduce-overhead')
cfn2 = torch.compile(torch.cos, mode='reduce-overhead')
for _ in range(2):
    x = cfn2(torch.randn(5).cuda())
y = cfn(x)
```

results in `Error: accessing tensor output of CUDAGraphs that has been overwritten by a subsequent run.`

But if I change the loop to `3`, there's no error. I think the error should not be there. Regardless, the behaviour should be the same whether the function has been run 1, 2, or 3+ times.

### Versions

Collecting environment information...
PyTorch version: 2.3.0a0+gitbcf7f9f
Is debug build: False
CUDA used to build PyTorch: 12.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (conda-forge gcc 12.3.0-2) 12.3.0
Clang version: Could not collect
CMake version: version 3.27.6
Libc version: glibc-2.35

Python version: 3.8.18 | packaged by conda-forge | (default, Oct 10 2023, 15:44:36)  [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-97-generic-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 2060
GPU 1: NVIDIA GeForce RTX 2060

Nvidia driver version: 545.23.08
cuDNN version: Probably one of the following:
/usr/local/cuda-11.7.1/targets/x86_64-linux/lib/libcudnn.so.8
/usr/local/cuda-11.7.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8
/usr/local/cuda-11.7.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8
/usr/local/cuda-11.7.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8
/usr/local/cuda-11.7.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8
/usr/local/cuda-11.7.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8
/usr/local/cuda-11.7.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8
/usr/local/cuda-12.3.2/targets/x86_64-linux/lib/libcudnn.so.9
/usr/local/cuda-12.3.2/targets/x86_64-linux/lib/libcudnn_adv.so.9
/usr/local/cuda-12.3.2/targets/x86_64-linux/lib/libcudnn_cnn.so.9
/usr/local/cuda-12.3.2/targets/x86_64-linux/lib/libcudnn_engines_precompiled.so.9
/usr/local/cuda-12.3.2/targets/x86_64-linux/lib/libcudnn_engines_runtime_compiled.so.9
/usr/local/cuda-12.3.2/targets/x86_64-linux/lib/libcudnn_graph.so.9
/usr/local/cuda-12.3.2/targets/x86_64-linux/lib/libcudnn_heuristic.so.9
/usr/local/cuda-12.3.2/targets/x86_64-linux/lib/libcudnn_ops.so.9
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      43 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             64
On-line CPU(s) list:                0-63
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen Threadripper 3970X 32-Core Processor
CPU family:                         23
Model:                              49
Thread(s) per core:                 2
Core(s) per socket:                 32
Socket(s):                          1
Stepping:                           0
Frequency boost:                    enabled
CPU max MHz:                        3700.0000
CPU min MHz:                        2200.0000
BogoMIPS:                           7400.38
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es
Virtualization:                     AMD-V
L1d cache:                          1 MiB (32 instances)
L1i cache:                          1 MiB (32 instances)
L2 cache:                           16 MiB (32 instances)
L3 cache:                           128 MiB (8 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-63
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec rstack overflow: Mitigation; safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] flake8==6.1.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.8.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.3
[pip3] onnx==1.15.0
[pip3] onnxruntime==1.17.0
[pip3] onnxscript==0.1.0.dev20240117
[pip3] optree==0.10.0
[pip3] pytorch-triton==3.0.0+989adb9a29
[pip3] torch==2.3.0a0+git93592f7
[pip3] torchvision==0.16.1+cf89794
[conda] libmagma                  2.7.2                h173bb3b_0    conda-forge
[conda] libmagma_sparse           2.7.2                h173bb3b_0    conda-forge
[conda] libopenvino-pytorch-frontend 2023.1.0             h59595ed_1    conda-forge
[conda] magma                     2.7.2                h51420fd_0    conda-forge
[conda] mkl                       2022.2.1         h84fe81f_16997    conda-forge
[conda] mkl-include               2023.2.0         h84fe81f_50495    conda-forge
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] optree                    0.10.0                   pypi_0    pypi
[conda] pytorch-triton            3.0.0+989adb9a29          pypi_0    pypi
[conda] torch                     2.3.0a0+git93592f7           dev_0    <develop>
[conda] torchfix                  0.4.0                    pypi_0    pypi
[conda] torchvision               0.16.1          cpu_py38h901811f_1    conda-forge

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184368

Make CUDAGraph replay outputs follow the same stale-output invalidation semantics as warmup and recording, including cached Tensor output eviction before poisoning stale storages.

Fixes #122192

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by bobrenjc93 — APPROVED** [review]

**Comment by zou3519** [comment]
@pytorchbot revert -c nosignal -m "really big regression in torchbench, see https://hud.pytorch.org/benchmark/v3/dashboard/compiler_inductor?renderGroupId=main&time.start=2026-05-15T00%3A00%3A00.000Z&time.end=2026-05-22T23%3A59%3A59.999Z&filters.repo=pytorch%2Fpytorch&filters.benchmarkName=compiler&filters.mode=inference&filters.dtype=bfloat16&filters.deviceName=cuda+%28h100%29&filters.device=cuda&filters.arch=h100&filters.suite=torchbench&filters.compiler=cudagraphs&lcommit.commit=f71f589ea2156794550aa485c676559b2052b0be&lcommit.workflow_id=25894735516&lcommit.date=2026-05-15T02%3A00%3A00Z&lcommit.branch=main&rcommit.commit=e88eee9588d216e0e0901751e34911981bc19df6&rcommit.workflow_id=26262699554&rcommit.date=2026-05-22T02%3A00%3A00Z&rcommit.branch=main&lbranch=main&rbranch=main&maxSampling=110"

**Comment by zou3519** [comment]
A codex bisect led me to this PR, codex says the following:
```
Bisect result:

first bad commit: 68f9ddc400239b62d69a4480695d45cd29dc61b7
Fix CUDAGraph stale output checks for cached tensors (#184368)

Key measurements with the CI-shaped command:

python benchmarks/dynamo/torchbench.py --performance --cold-start-latency \
  --device cuda --inference --bfloat16 --backend inductor \
  --only lennard_jones -n50

Results:

- b298938ef64f parent-side good: 1.329718x
- 68f9ddc40023 first bad: 0.969288x
- d0 / 04d232f35374: 0.921x to 0.955x
- d0 --disable-cudagraphs: 0.871x, basically unchanged from earlier no-cg runs

The regression is not from losing capture: all bad/good runs still show calls_captured=9, unique_graphs=1, graph_breaks=0,
cudagraph_skips=0.

The culprit adds execution-ending logic in torch/_inductor/cudagraph_trees.py that can checkpoint allocator state, drop cached replay
tensors, and poison/free stale storages at generation boundaries. For a tiny model like lennard_jones, that cudagraph tree cleanup/
reconstruction overhead appears to dominate runtime.

```

**Review by eellison — COMMENTED** [review]
@jansel please make sure to include code owners in request for review for cudagraphs

**Comment by jansel** [comment]
My agent says I requested @eellison for the CUDAGraphs code-owner review and updated the PR after the revert. The revised fix keeps cached replay tensors on ordinary fast-path rollover and only poisons live cached replay outputs when leaving execution to warm up or record, which addresses the TorchBench lennard_jones regression while preserving the stale-output check.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I reran the failed ROCm distributed CI jobs. The logs point to ROCm distributed failures and worker-side SkipTest handling outside this CUDAGraphs change, so I will revisit once the rerun finishes.

**Comment by jansel** [comment]
My agent says I updated the ghstack PR after rebasing and fixed an additional stale-output path caught during review: same-root replay keeps cached tensors, while switching between already-recorded roots now poisons live cached outputs before replay. Focused CUDAGraph tests and lintrunner -a passed.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I reran the failed CI jobs on the current head. The failures I checked were outside this CUDAGraphs change or CI infrastructure/artifact-upload issues, so I will revisit after the reruns finish.

**Comment by jansel** [comment]
My agent says I rebased and resubmitted the PR on current main after checking the failed CI. The rerun artifacts/logs did not show a CUDAGraphs-specific failure; CUDA cooperative reductions passed on rerun, ROCm showed an unrelated compiled-autograd expect-test mismatch, and the remaining failures were broken-trunk/infra. Focused CUDAGraph tests and lintrunner passed; CI is pending on the refreshed head.

**Comment by jansel** [comment]
My agent says I rebased and resubmitted the PR on current main, and fixed an additional stale-output case found during review: same-root successful replay still preserves cached Tensor outputs, but same-root re-record after static input pointer churn now poisons stale cached outputs before recording a fresh root. Focused CUDAGraph tests and lintrunner passed.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I inspected the current Dr. CI failure. The ROCm `test_cuda_graph_utils.py::TestMarkKernels::test_clear_resets_state` job failed on all retries with `cudaErrorInsufficientDriver`, which points to the ROCm runner/driver environment rather than this CUDAGraphs tree change; the TorchTitan failure is marked broken trunk by Dr. CI. I reran the failed ROCm workflow and will revisit after it completes.

**Comment by jansel** [comment]
Agent, has the regression that caused the prior revert been fixed?  Share some benchmark numbers.

**Comment by jansel** [comment]
My agent says yes, the regression that caused the revert has been addressed by keeping same-root CUDAGraph replay on the cheap cached-output path and only poisoning/deallocating live outputs when execution transitions to warmup, recording, a different root, or a same-root re-record. The revert report measured the original landed PR at 0.969288x on `python benchmarks/dynamo/torchbench.py --performance --cold-start-latency --device cuda --inference --bfloat16 --backend inductor --only lennard_jones -n50`, versus 1.329718x on the parent-side good commit. After the revised fix, I ran the same command locally and got 1.351x, which is back in line with the pre-regression number while preserving the stale-output checks.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Comment by jansel** [comment]
My agent says the latest mergebot attempt was blocked only by the ROCm `test_cuda_graph_utils.py::TestMarkKernels::test_clear_resets_state` job that previously failed with `cudaErrorInsufficientDriver`, while Dr. CI marks the TorchTitan failure as broken trunk. I reran the failed ROCm workflow and will revisit after it completes.

**Comment by jansel** [comment]
My agent says I checked the completed ROCm rerun and it is still the same runner/driver failure: `test_cuda_graph_utils.py::TestMarkKernels::test_clear_resets_state` raises `cudaErrorInsufficientDriver` before this PR's CUDAGraph tree code is involved. I reran the failed workflow again; attempt 4 is queued, so I will revisit after it completes.

**Comment by jansel** [comment]
My agent says I checked the completed ROCm rerun and it is still the same runner/driver failure: test_cuda_graph_utils.py::TestMarkKernels::test_clear_resets_state raises cudaErrorInsufficientDriver before this PR's CUDAGraph tree code is involved. I reran the failed workflow again and will revisit after it completes.

**Comment by jansel** [comment]
My agent says I checked the latest ROCm rerun artifact for attempt 5 and it is still the same runner/driver failure: test_cuda_graph_utils.py::TestMarkKernels::test_clear_resets_state raises cudaErrorInsufficientDriver before this PR's CUDAGraph tree code is involved. I reran the failed workflow again; attempt 6 is queued, so I will revisit after it completes.

**Comment by jansel** [comment]
My agent says I checked the completed ROCm rerun attempt 6 and it is still the same runner/driver failure: `test_cuda_graph_utils.py::TestMarkKernels::test_clear_resets_state` raises `cudaErrorInsufficientDriver` before this PR’s CUDAGraph tree code is involved. I reran the failed trunk workflow again and will revisit after it completes.

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main to pick up the upstream ROCm TestMarkKernels skip for the repeated cudaErrorInsufficientDriver failure. While resolving the CUDAGraph tree conflict, I kept the stale cached-output poisoning behavior and adjusted same-root successful replay to execute directly from the precheck, avoiding an extra invariant check/list copy on the hot path. Syntax checks and lintrunner with Pyrefly skipped passed locally; CI is pending on the refreshed head.

**Review by eellison — CHANGES_REQUESTED** [review]
marking this issue as no agent

**Comment by jansel** [comment]
My agent says this PR is blocked by the issue being marked no-agent / agents-banned, with changes requested for that reason. I am closing this agent-created PR rather than continuing to push or request review.
